File: app/storage.py
"""状态持久化存储模块。

负责从 data/state.json 读取和保存应用状态，保证线程/协程安全。
"""

import asyncio
import json
import os
from typing import Any, Dict

from app.auth import get_default_password, hash_password
from app.models import AppState, NotifySettings, ScheduleSettings, SecuritySettings, Settings


class StateManager:
    """系统状态管理器，提供线程/协程安全的 JSON 文件读写功能。"""

    def __init__(self, state_path: str = os.path.join("data", "state.json")) -> None:
        self.state_path: str = state_path
        self._lock: asyncio.Lock = asyncio.Lock()
        self.state: AppState = self._default_state()

    def _default_state(self) -> AppState:
        """生成系统默认配置状态。"""
        return AppState(
            next_task_id=1,
            next_history_id=1,
            tasks=[],
            history=[],
            settings=Settings(
                notify=NotifySettings(notify_on_failure=True),
                schedule=ScheduleSettings(hour=8, minute=0, second=0, max_workers=4),
                security=SecuritySettings(
                    enabled=True,
                    password_hash=hash_password(get_default_password()),
                ),
            ),
        )

    async def load(self) -> None:
        """异步加载 data/state.json 配置文件。如果文件不存在或损坏则使用默认状态。"""
        async with self._lock:
            if not os.path.exists(self.state_path):
                os.makedirs(os.path.dirname(self.state_path), exist_ok=True)
                self.state = self._default_state()
                await self._save_unlocked()
                return

            try:
                with open(self.state_path, "r", encoding="utf-8") as f:
                    raw_data: Dict[str, Any] = json.load(f)
                state_obj = AppState.model_validate(raw_data)
            except Exception:
                state_obj = self._default_state()

            # 兼容性容错修复
            if state_obj.next_task_id < 1:
                state_obj.next_task_id = 1
            if state_obj.next_history_id < 1:
                state_obj.next_history_id = 1
            if not state_obj.settings.security.password_hash:
                state_obj.settings.security.password_hash = hash_password(get_default_password())
                state_obj.settings.security.enabled = True

            self.state = state_obj

    async def save(self) -> None:
        """保存状态到磁盘（加锁）。"""
        async with self._lock:
            await self._save_unlocked()

    async def _save_unlocked(self) -> None:
        """写入临时文件后重命名以实现原子写入。"""
        os.makedirs(os.path.dirname(self.state_path), exist_ok=True)
        tmp_path: str = f"{self.state_path}.tmp"
        json_data: str = self.state.model_dump_json(indent=2)
        with open(tmp_path, "w", encoding="utf-8") as f:
            f.write(json_data)
        os.replace(tmp_path, self.state_path)

File: storage.go
package main

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sync"
)

// StateManager 状态管理器，提供线程安全的 JSON 文件持久化
type StateManager struct {
	StatePath string
	Mu        sync.RWMutex
	State     AppState
}

// NewStateManager 创建状态管理器实例
func NewStateManager(path string) *StateManager {
	if path == "" {
		path = filepath.Join("data", "state.json")
	}
	return &StateManager{
		StatePath: path,
		State:     defaultState(),
	}
}

func defaultState() AppState {
	return AppState{
		NextTaskID:    1,
		NextHistoryID: 1,
		Tasks:         []Task{},
		History:       []HistoryItem{},
		Settings: Settings{
			Notify: NotifySettings{
				NotifyOnFailure: true,
			},
			Schedule: ScheduleSettings{
				Hour:       8,
				Minute:     0,
				Second:     0,
				MaxWorkers: 4,
			},
			Security: SecuritySettings{
				Enabled:      true,
				PasswordHash: HashPassword(GetDefaultPassword()),
			},
		},
	}
}

// Load 加载 state.json 配置文件
func (sm *StateManager) Load() error {
	sm.Mu.Lock()
	defer sm.Mu.Unlock()

	if _, err := os.Stat(sm.StatePath); os.IsNotExist(err) {
		if err := os.MkdirAll(filepath.Dir(sm.StatePath), 0755); err != nil {
			return fmt.Errorf("创建数据目录失败: %w", err)
		}
		sm.State = defaultState()
		return sm.saveUnlocked()
	}

	data, err := os.ReadFile(sm.StatePath)
	if err != nil {
		slog.Error("读取状态文件失败，恢复默认配置", "error", err)
		sm.State = defaultState()
		return sm.saveUnlocked()
	}

	var loaded AppState
	if err := json.Unmarshal(data, &loaded); err != nil {
		slog.Error("解析状态文件 JSON 失败，恢复默认配置", "error", err)
		sm.State = defaultState()
		return sm.saveUnlocked()
	}

	// 容错与修复
	if loaded.NextTaskID < 1 {
		loaded.NextTaskID = 1
	}
	if loaded.NextHistoryID < 1 {
		loaded.NextHistoryID = 1
	}
	if loaded.Tasks == nil {
		loaded.Tasks = []Task{}
	}
	if loaded.History == nil {
		loaded.History = []HistoryItem{}
	}
	if loaded.Settings.Security.PasswordHash == "" {
		loaded.Settings.Security.PasswordHash = HashPassword(GetDefaultPassword())
		loaded.Settings.Security.Enabled = true
	}

	sm.State = loaded
	return nil
}

// Save 安全保存状态到文件（自动加写锁）
func (sm *StateManager) Save() error {
	sm.Mu.Lock()
	defer sm.Mu.Unlock()
	return sm.saveUnlocked()
}

// saveUnlocked 写入临时文件后原子重命名
func (sm *StateManager) saveUnlocked() error {
	if err := os.MkdirAll(filepath.Dir(sm.StatePath), 0755); err != nil {
		return fmt.Errorf("创建数据目录失败: %w", err)
	}

	data, err := json.MarshalIndent(sm.State, "", "  ")
	if err != nil {
		return fmt.Errorf("序列化状态数据失败: %w", err)
	}

	tmpPath := sm.StatePath + ".tmp"
	if err := os.WriteFile(tmpPath, data, 0644); err != nil {
		return fmt.Errorf("写入临时状态文件失败: %w", err)
	}

	if err := os.Rename(tmpPath, sm.StatePath); err != nil {
		return fmt.Errorf("覆盖替换状态文件失败: %w", err)
	}
	return nil
}
